### DISCUSSION TITLE
RFC: Cloudflare Data Hooks

### DISCUSSION THREAD

**Discussion by jplhomer** [flareact/flareact#49] [Ideas]
## Background

Cloudflare Workers provides access to a [KV Store](https://developers.cloudflare.com/workers/learning/how-kv-works) for simple globally-distributed storage. Cloudflare also recently announced support for [Durable Objects (limited beta)](https://developers.cloudflare.com/workers/learning/using-durable-objects) within Workers for more realtime, consistent and available storage.

Both of these require some effort on a developers part to:

1. Create the namespace
1. Connect the namespace to the Workers project
1. Recall and instantiate the namespace to use the API

With the simplicity Flareact already provides to create an edge-rendered React app, wouldn't it be great if tapping into either of these data storage mechanisms were as simple as a React hook?

## Proposal

Flareact should provide an API to **create namespaces**, **read** and **write** to first-party data storage mechanisms (i.e. those provided by Cloudflare). Flareact will provide **default namespaces**, meaning the developer doesn't even need to think about what to name their first namespace. It will also provide the ability to create and map additional non-default namespaces.

```bash
yarn create flareact-app my-app

Creating new Flareact app....

> Would you like to create a default KV namespace? (y/N) y
> Would you like to create a default Durable Objects namespace? (y/N) y
```

You could also imagine a shorthand property to create a **default KV namespace AND Durable Objects namespace** in one line:

```bash
yarn create flareact-app my-app --kv --object
```

## KV Store

At a high level, the React API for **KV** might look like this:

```js
export default function Posts() {
  const [posts] = useData();

  if (!posts) return <p>Loading...</p>;

  return posts.map(() => ...);
}
```

```js
export default function SinglePost() {
  const [post, setPost] = useData('post-12345');

  if (!post) return <p>Loading...</p>;

  return <h1>post.title</h1>;
}
```

```js
export default function AddPost() {
  const [posts, addPost] = useData();

  async function handleSubmit(e) {
    e.preventDefault();

    await addPost({ ... });
  }

  return <form onSubmit={handleSubmit}>...</form>;
}
```

### Alternate Implementations

One pitfall to the array-based syntax is that for e.g. write requests, we need to send the `set` parameter as the second value. This means all keys will be fetched and listed from the store, even if the user only wishes to add a new key value.

An alternative to the array-based hook above would be to use an object-based hook:

```js
export default function Posts() {
  const { data: posts, loading, error } = useData();

  if (loading) return <p>Loading...</p>;

  return posts.map(() => ...);
}
```

We would then need to define a slightly different hook (think `useQuery` and `useMutation` within Apollo Client) to send write requests:

```js
export default function AddPosts() {
  const [addPost, { loading, error }] = useDataMutation(); 

  async function handleSubmit(e) {
    e.preventDefault();

    await addPost({ ... });
  }

  return <form onSubmit={handleSubmit}>...</form>;
}
```

### Edge and utility access

We need a way to abstract the `NAMESPACE` constant into a Flareact utility, which will also be used within the API request.

Let's follow the pattern Cloudflare is taking with their modules-based Workers migration by passing an `env` object to user-defined methods like `getEdgeProps` and any API requests:

```js
// pages/index.js
export async function getEdgeProps({ env }) {}

// pages/api/hello.js
export async default (event, env) {}
```

By providing a Flareact utility, the user needs only to pass the provided `env` to get a nice API for interacting with the store - which closely resembles the already-nice API present in KV & Durable Objects:

```js
import { getKVStore } from 'flareact/data';

export async function getEdgeProps({ env }) {
  const store = getKVStore(env);

  const posts = await store.list();

  // e.g.
  // const post = await store.get('post-1234');

  return {
    props: { posts.keys }
  };
}
```

### Implementation Details

The client React hooks are modeled after [SWR](https://swr.vercel.app/) and the standard [useState](https://reactjs.org/docs/hooks-state.html) hook.

These would **proxy requests to a framework-level Flareact API endpoint**. 

When the hook is called, a Fetch request is made behind the scenes:

```
/_flareact/_store/kv/read
```

A (internal, framework-only) endpoint handler will be defined with Flareact to fetch the requested data and return a response:

```js
import { getKVStore } from 'flareact/data';

export default async (event, env) => {
  const params = await event.request.json();
  const store = getKVStore(env);

  return (await store.list(params)).keys;
} 
```

- Instead of `NAMESPACE`, Flareact will inject the namespace a user has defined (likely through a CLI command or setup process) which is printed in `wrangler.toml`.

For **write requests**, it's likely that the developer will want to limit access to certain users. We can accomplish this by allowing the user to define a `write` policy within `policies/kv.js`:

```js
import userIsValid from '../some/auth/library.js';

export async write(event) {
  const { Authorization: token } = event.request.headers;

  return await userIsValid(token);
}
```

This policy mechanism could also extend to `read`, `delete`, `list`, and even blanket `access` actions within the same file.

Notes on this:

- We probably want to reject any `write` actions with a default policy until the user has explicitly defined one.

## Durable Objects

Durable Objects are essentially instances of JavaScript classes that can be fetched on a global basis. The data is guaranteed to be consistent and up-to-date.

We'll offer a simple `useRealtimeData` along with a default Durable Object class to get the user started:

```js
export default function ChatRoom({ roomId }) {
  const { data, emit } = useRealtimeData(roomId);
  const { messages, participants } = data;

  async function handleSubmit(e) {
    emit('newMessage', { ... });
  }

  <form onSubmit={handleSubmit}>...</form>
}
```

Under the hood, Flareact handles creating a **WebSocket**, listening for new data, and emitting data back to the WebSocket.

It's likely the developer will want to define a custom Durable Object class. They can create new durable objects in their Flareact project:

```js
// objects/room.js

export class Room {
  async fetch(event) {
    // ...
  }
}
```

And reference them with the hook:

```diff
export default function ChatRoom({ roomId }) {
-  const { data, emit } = useRealtimeData(roomId);
+  const { data, emit } = useRealtimeData('room', roomId);
  const { messages, participants } = data;
}
```

Caveats and notes:

- Still early days for the Durable Objects beta (I don't have access) and stuff may change
- May require larger architectural shifts for Flareact to be able to support `mjs` workers

## More fun ideas

Since both KV and Durable Objects provide a simple KV store, you can build your own schema using smart prefixes, e.g. `user:1`, `post:2`, etc. 

At the same time, it's likely a user will want to attach `metadata` to a given KV so it is available with the `NAMESPACE.list()` call. _Imagine wanting to list out all posts with a title + ID, but use the ID as the primary key._

We could provide a **model-based abstraction** around this concept. Similar to Rails' ActiveModel and Laravel's Eloquent, the user could create a file for a given model within their project:

```js
// models/post.js

export default class Post extends Model {
  excerpt() {
    return this.body.slice(0, 100);
  }
}
```

The developer would then reference that custom model when writing and accessing the data. 

When Flareact **writes** data, it will store it using a prefix derived automatically from that JavaScript class name. It will also provide nice things like automatic UUID generation. For the example above: `post:<UUID>`.

When Flareact **reads** data, it will instantiate the custom model with the data from the KV store. This allows the user to define custom getters and methods _on top_ of their existing data. It also opens the door to doing some really cool stuff with relations between models, e.g. `await post.user()`.

This model can also intelligently store properties inside `metadata` so they are available with `list()` calls.

The extended class provides all the hooks and interfaces necessary for Flareact to perform what it needs to do:

```js
import { useData} from '../models/post.js';

export default function Home() {
  const { posts, loading, error } = useData();

  return posts.map(post => (
    <div>
      <h2>{post.title}</h2>
      <p>{post.excerpt()}</p>
    </div>
  ));
}
```

**Comment by banderson623** [comment]
+1

**Comment by bpernick** [comment]
I like this idea and would be happy to take a stab at implementing if people don't mind!

**Comment by jplhomer** [comment]
@bpernick That's terrific news ❤️ I got a head-start here: https://github.com/flareact/flareact/compare/cf-data-hooks Would love to have you contribute! 

Next steps are:

- Introduce the concept of `policies`
- Support KV mutations
- Add `flareact kv` commands for creating a default namespace

**Comment by bpernick** [comment]
Great @jplhomer! I can start looking at that tonight :)

**Comment by jplhomer** [comment]
Been thinking about some concerns around this recently:

## `list` call rate limits

IIRC Cloudflare limits you to 1,000 `list` KV calls per day. This is pretty restrictive and an indicator that this method should be used sparingly.

While it would be awesome to have a React hook for listing out all keys in the store, I think this is unrealistic to support given this limitation. I'm leaning toward cutting the list output from the client-side hook and only supporting hook references to specific keys, e.g. `useData(<my-key-here>)` as the limitations are much more lax on those.

## General security concerns

My original intention was to provide `read` access via the hooks open to the world by default. This provides a dev experience with the least friction. However, it offers carte blanche access to the developer's entire namespace. 

Without smartly defining a `read` access policy, I'm worried the `/_flareact/data/kv/` API endpoint used behind the scenes will serve as a vector for attack. Whether it's to exhaust a user's CF rate limit, discover keys the developer hadn't intended to make public, etc.

We should consider changing the "open by default" policy to something else... or just not shipping the client-side hook feature just yet 😬 

The edge-side utilities are still great and wonderful for DX - plus, the developer explicitly makes a choice to share data at the edge.

**Reply by bpernick** [reply]
Hey Josh I see your point about security -- allowing random people to fish around your kv store doesn't sound ideal. I have not worked with Cloudflare security policies and am having a little trouble finding decent information; if you get a chance would you mind sending over a few links?

**Reply by bpernick** [reply]
I agree adding a list hook is probably not worth it

**Reply by bpernick** [reply]
As for rate limit attacks, would this not be a concern even without read and write hooks?

**Reply by jplhomer** [reply]
As far as policies are concerned, it would all be contained within Flareact, where the developer has the ability to define e.g. `read` policies based on an incoming request. I started in on this here: https://github.com/flareact/flareact/compare/cf-data-hooks but still not settled on whether I want to extend the policy functionality to the edge utilities in addition to hooks, or just hooks.

Rate limit attacks should be mitigated at the edge if the user sets a cache on the `revalidate` property of `getEdgeProps` (or manually in an API endpoint).

**Reply by traffisco** [reply]
FYI,
Beyond the free tier, KV operations are very reasonably priced. 
`Additional KV usage costs $0.50 per million read operations, $5.00 per million list, write and delete operations and $0.50 per GB of stored data.`

**Reply by bpernick** [reply]
@jplhomer what do you think about creating a default read policy that allow the dev to either blacklist or whitelist certain keys?

**Reply by bpernick** [reply]
Here's a quick example of the kind of thing I was thinking, using env variables for BLACKLIST_MODE and the black and white lists

**Reply by bpernick** [reply]
const blacklist = BLACKLIST.split(',');
const whitelist = WHITELIST.split(',');

const DEFAULT_READ_CLIENT_POLICY = (request) => {
  const url = new URL(request.url);
  const params = url.searchParams;
  const key = params.get("key");
  
  if (BLACKLIST_MODE) {
    if (blacklist.includes(key)) {
      return false;
    }
    return true;
  } else {
    if (!whitelist.includes(key)) {
      return false;
    }
    return true;
  }
};
